tket/src/WeightSubgrMono/EndToEndWrappers/MainSolver.cpp

```diff
@@ -0,0 +1,421 @@
+          return;
+        }
+        // Make it strictly better.
+        --reduction_parameters.max_weight;
+        reduction_parameters.max_weight = std::min(
+            reduction_parameters.max_weight, initial_weight_upper_bound);
+      }
+    }
+
+    if (reduction_parameters.max_weight <
+        m_solution_data.trivial_weight_lower_bound) {
+      m_solution_data.finished = true;
+      return;
+    }
+
+    // Now we can search!
+    ++m_solution_data.iterations;
+
+    // On the first move ONLY, we don't backtrack; but we also haven't reduced.
+    if (m_solution_data.iterations == 1) {
+      if (!m_search_branch_ptr->reduce_current_node(reduction_parameters)) {
+        m_solution_data.finished = true;
+        return;
+      }
+    } else {
+      if (!m_search_branch_ptr->backtrack(reduction_parameters)) {
+        m_solution_data.finished = true;
+        return;
+      }
+    }
+    if (move_down_from_reduced_node(reduction_parameters)) {
+      // We've GOT a complete solution! Note that it MUST be good enough
+      // to add, since we've set the max weight already.
+      // We also already checked that we haven't yet got too many,
+      // if we're storing more than one.
+      add_solution_from_final_node(parameters, reduction_parameters);
+      if (terminate_with_enough_full_solutions(parameters, m_solution_data)) {
+        return;
+      }
+    }
+    if (Clock::now() >= desired_end_time) {
+      return;
+    }
+  }
+}
+
+void MainSolver::add_solution_from_final_node(
+    const MainSolverParameters& parameters,
+    const SearchBranch::ReductionParameters& reduction_parameters) {
+  TKET_ASSERT(m_pre_search_components_ptr);
+  TKET_ASSERT(m_search_branch_ptr);
+
+  const auto& accessor = m_search_branch_ptr->get_domains_accessor();
+  const auto scalar_product = accessor.get_scalar_product();
+
+  TKET_ASSERT(
+      accessor.get_total_p_edge_weights() ==
+      m_solution_data.total_p_edge_weights);
+
+  TKET_ASSERT(scalar_product <= reduction_parameters.max_weight);
+
+  // We'll overwrite the solution into back().
+  if (parameters.for_multiple_full_solutions_the_max_number_to_obtain > 0 ||
+      m_solution_data.solutions.empty()) {
+    m_solution_data.solutions.emplace_back();
+  }
+  auto& assignments = m_solution_data.solutions.back().assignments;
+  const auto number_of_pv = accessor.get_number_of_pattern_vertices();
+  assignments.clear();
+  assignments.reserve(number_of_pv);
+
+  for (unsigned pv = 0; pv < number_of_pv; ++pv) {
+    const auto& domain = accessor.get_domain(pv);
+    TKET_ASSERT(domain.size() == 1);
+    assignments.emplace_back(pv, *domain.cbegin());
+  }
+  m_solution_data.solutions.back().scalar_product = scalar_product;
+
+  m_solution_data.solutions.back().total_p_edges_weight =
+      m_solution_data.total_p_edge_weights;
+}
+
+bool MainSolver::move_down_from_reduced_node(
+    const SearchBranch::ReductionParameters& reduction_parameters) {
+  TKET_ASSERT(m_search_components_ptr);
```

Comment:
Actually, in this case, strictly speaking NO: m_search_components_ptr is initialised only if "initialisation" succeeded in the constructor; otherwise it is null.

You are right that this is a private function which should be impossible to call UNLESS "initialisation" succeeded; but again, asserts are only SUPPOSED to be for things which are "impossible to fail".

Reply:
"bar hardware failure, that ptr should always exist": 

No, the ptr can still be null, if it detects that the problem is insoluble in the constructor, where it is normally set.

In that case, calling move_down_from_reduced_node would be a programming error.

OK, it's a private function, so the possible callers are restricted; it would require someone to change the code and introduce a bug. However I still think it's right to assert. I don't think we should distinguish between public and private functions for this purpose (after all, public functions can call private functions; who can tell the possible chains of calls?)

The whole POINT of an assert is to avoid spending ages thinking about whether something could possibly be called in an unusual way due to a programming error. By worrying about what could cause an assert failure, I really feel you're missing the point!

Suppose that every one of these asserts you don't like is deleted. Then 5 years from now, or even after both of us are dead, if something goes wrong which WOULD have been caught if the assert were still there, then time and money is wasted due to more difficult debugging (or, even worse, nonsense unreproducible results are spewed out without warning): (1) the value of an assert could last way into the future; (2) no matter how obvious, you can never be sure WHICH asserts are worth keeping. You might as well say, "don't bother checking everything; just check the stuff which won't pass" !

Reply:
"If you had another class method that deallocated the ptr manually then this TKET_ASSERT could be reasonable ..."

You're not suggesting we maintain a record, for all time, of EVERY one of these asserts you want to delete, AND check the record after EVERY code change, and keep it up to date - adding/deleting/re-adding asserts as appropriate??!

Reply:
No I'm not suggesting that. 

I suspect we are not going to come to an agreement over where TKET_ASSERT should be used - I will resolve the conversations.